org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/GetInferSelectionHandler.java

```diff
@@ -0,0 +1,96 @@
+
+package org.eclipse.jdt.ls.core.internal.handlers;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.dom.ASTNode;
+import org.eclipse.jdt.core.dom.Expression;
+import org.eclipse.jdt.core.dom.ParenthesizedExpression;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.corext.refactoring.code.ExtractMethodRefactoring;
+import org.eclipse.jdt.ls.core.internal.corrections.DiagnosticsHelper;
+import org.eclipse.jdt.ls.core.internal.corrections.InnovationContext;
+import org.eclipse.jdt.ls.core.internal.text.correction.RefactorProposalUtility;
+import org.eclipse.lsp4j.CodeActionParams;
+import org.eclipse.lsp4j.FormattingOptions;
+
+public class GetInferSelectionHandler {
+
+	public static List<InferSelection> getInferSelectionsForRefactor(GetInferSelectionParams params) {
+		final ICompilationUnit unit = JDTUtils.resolveCompilationUnit(params.context.getTextDocument().getUri());
+		if (unit == null) {
+			return null;
+		}
+		Map formatterOptions = params.options == null ? null : FormatterHandler.getOptions(params.options, unit);
+		int start = DiagnosticsHelper.getStartOffset(unit, params.context.getRange());
+		int end = DiagnosticsHelper.getEndOffset(unit, params.context.getRange());
+		InnovationContext context = new InnovationContext(unit, start, end - start);
+		List<InferSelection> extractMethodInfos = new ArrayList<InferSelection>();
+		ASTNode parent = context.getCoveringNode();
+		try {
+			if (RefactorProposalUtility.EXTRACT_METHOD_COMMAND.equals(params.command)) {
+				while (parent != null && parent instanceof Expression) {
+					if (parent instanceof ParenthesizedExpression) {
+						parent = parent.getParent();
+						continue;
+					}
+					ExtractMethodRefactoring refactoring = new ExtractMethodRefactoring(context.getASTRoot(), parent.getStartPosition(), parent.getLength(), formatterOptions);
+					if (refactoring.checkInitialConditions(new NullProgressMonitor()).isOK()) {
+						extractMethodInfos.add(new InferSelection(parent.toString(), parent.getStartPosition(), parent.getLength()));
+					}
+					parent = parent.getParent();
+				}
+			}
+
+		} catch (CoreException e) {
+			// do nothing.
+		}
+
+		if (extractMethodInfos.size() == 0) {
+			return null;
+		}
+		return extractMethodInfos;
+	}
+
+	public static class InferSelection {
+		public String name;
+		public int startPosition;
+		public int length;
+
+		public InferSelection(String name, int startPosition, int length) {
+			this.name = name;
+			this.startPosition = startPosition;
+			this.length = length;
+		}
+	}
+
+	public static class GetInferSelectionParams {
+		public String command;
+		public CodeActionParams context;
+		public FormattingOptions options;
```

Comment:
When inferring a possible expression for the refactoring operation, formatting options is not needed.

Reply:
Will remove this unnecessary field.